File: solutions/clojure/sublist/3/src/sublist.clj
(ns sublist)

(defn starts-with? [list prefix]
  (cond
    (empty? prefix) true
    (empty? list) false
    (= (first list) (first prefix)) (recur (rest list) (rest prefix))
    :else false))

(defn- sublist?
  "Whether the l1 is a sublist of l2"
  [l1, l2]
  (cond
    (empty? l2) false
    :else (if (starts-with? l2 l1)
            true
            (recur l1 (rest l2)))))

(defn classify
  "Returns:
  :equal if coll1 equals coll2,
  :superlist if coll1 is a superlist of coll2,
  :sublist if coll1 is a sublist of coll2,

  If none of these conditions is true, it returns :unequal."
  [coll1 coll2]
  (cond
    (= coll1 coll2) :equal
    (and (< (count coll1) (count coll2)) (sublist? coll1 coll2)) :sublist
    (and (> (count coll1) (count coll2)) (sublist? coll2 coll1)) :superlist
    :else :unequal))

File: solutions/elixir/sublist/3/lib/sublist.ex
defmodule Sublist do
  @doc """
  Returns whether the first list is a sublist or a superlist of the second list
  and if not whether it is equal or unequal to the second list.
  """
  def compare(a, b) do
    cond do
      a == b -> :equal
      a == [] -> :sublist
      b == [] -> :superlist
      true -> do_compare(a, b)
    end
  end

  defp do_compare(a, b) do
    b_chunks = Stream.chunk_every(b, length(a), 1, :discard)

    if Enum.any?(b_chunks, &(&1 === a)) do
      :sublist
    else
      a_chunks = Stream.chunk_every(a, length(b), 1, :discard)

      if Enum.any?(a_chunks, &(&1 === b)) do
        :superlist
      else
        :unequal
      end
    end
  end
end
